### PR TITLE
display charts with data computed by backend when form is submitted

### PR DESCRIPTION
**Pull request by krachwal**
Updated placeholder copy to copy received from Michael and wired Analyze button to render charts based on form data instead of hardcoded data when user clicks the Analyze button.

Also:
- Broke Python main.py into separate files
- Set up backend logging
- Upgraded Cypress
- Added Cypress e2e test for mixed variables validation
- Added Cypress e2e test for Analyze button click transition to Charts component

New form copy:

image

New validation of single variable use in each form field:

image

New validation of single variable use across form:

image

New charts component with real data based on inputs and updated copy:

image

### CODE REVIEW

File: python-backend/math_utils.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+from sympy import Symbol
+
+import constants
+
+GRAPHABLE_TYPES = (int, float, sympy.core.numbers.Integer, sympy.core.numbers.Float)
+PRECISION = constants.PRECISION
+X_VALUES = range(0, 1000, 25)
+
+
+def error_coordinates(expression: sympy.core, symbol: Symbol, limit: sympy.core.numbers):
```
Comment (davidbrownell) [inline python-backend/math_utils.py:11]: Provide the return type. Please do this for all python functions that you write.

File: python-backend/math_utils.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    """
+    graph coords for the error of an expression log(|expression - L|) in terms of symbol
+    :param expression: univariate mathematical expression comprising a ratio of two polynomials
+    :param symbol: the variable used in expression
+    :param limit: the computed limit of the expression as it goes to infinity
+    :return: array of [x,y] pairs for graphing purposes
+    """
+    return [[val, float(sympy.log(abs(expression.subs(symbol, val) - limit)).evalf(PRECISION))] for val in X_VALUES if
```
Comment (davidbrownell) [inline python-backend/math_utils.py:19]: This return value is pretty complicated, so I tried to add indentation to make it more legible but I'm not sure that I got it right.

```
[
    [
        val, 
        float(sympy.log(abs(expression.subs(symbol, val) - limit)).evalf(PRECISION))
    ]
    for val in X_VALUES 
    if type(sympy.log(abs(expression.subs(symbol, val) - limit)).evalf(PRECISION)) in GRAPHABLE_TYPES
]
```

Consider creating a `dataclass`-based type to replace the inner list.

File: python-backend/math_utils.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    """
+    graph coords for the error delta for an expression -1 * (log(|Pn/Qn - L|) / log(Qn)) - 1 in terms of symbol
+    :param expression: univariate mathematical expression comprising a ratio of two polynomials
+    :param denominator: univariate polynomial
+    :param symbol: the variable used in expression
+    :param limit: the computed limit of the expression as it goes to infinity
+    :return: array of [x,y] pairs for graphing purposes
+    """
+    # graph coords of error delta: -1 * (log(|Pn/Qn - L|) / log(Qn)) - 1
+    return [[val, float(-1 * ((sympy.log(abs(expression.subs(symbol, val) - limit)).evalf(PRECISION)) / (
```
Comment (davidbrownell) [inline python-backend/math_utils.py:33]: Same comments as above.

File: python-backend/wolfram_client.py [hunk 4]
```diff
@@ -29,4 +36,18 @@ def ask(query: str) -> str:
 
     @staticmethod
     def limit(fraction: str) -> str:
-        return WolframClient.ask(query="limit of {}".format(fraction))
+        try:
+            result = WolframClient.ask(query="limit of {}".format(fraction), include_pod="Limit")
+            # only want to return: queryresult -> pods[0] -> subpods
+            return result["queryresult"]["pods"][0]["subpods"]
+        except Exception as e:
+            print("Failed to parse Wolfram API result", e)
```
Comment (davidbrownell) [inline python-backend/wolfram_client.py:44]: Should this use `logging` rather than `print`?

File: python-backend/wolfram_client.py [hunk 5]
```diff
@@ -29,4 +36,18 @@ def ask(query: str) -> str:
...
+
+    @staticmethod
+    def continued_fraction(fraction: str) -> str:
+        try:
+            result = WolframClient.ask(query="continued fraction representation of {}".format(fraction))
+            # only want to return: queryresult -> pods[0] -> subpods
+            return result["queryresult"]["pods"][0]["subpods"]
+        except Exception as e:
+            print("Failed to parse Wolfram API result", e)
```
Comment (davidbrownell) [inline python-backend/wolfram_client.py:53]: Same comment as above.

**Review by davidbrownell — CHANGES_REQUESTED** [review]

File: python-backend/math_utils.py (hunk 1 above)
Comment (krachwal) [reply python-backend/math_utils.py:11]: I always forget. I'll see if I can enable a check to remind me to do this...

File: python-backend/wolfram_client.py (hunk 4 above)
Comment (krachwal) [reply python-backend/wolfram_client.py:44]: great catch!

File: python-backend/math_utils.py (hunk 2 above)
Comment (krachwal) [reply python-backend/math_utils.py:19]: I agree that it's illegible. It's just a list of `int, float` pairs so I specified that and broke out the nested logic for legibility.

**Review by davidbrownell — APPROVED** [review]

**Pull request merged**